service/history/mutableStateBuilder.go

```diff
@@ -119,12 +125,93 @@ func (e *mutableStateBuilder) Load(state *persistence.WorkflowMutableState) {
 	e.pendingChildExecutionInfoIDs = state.ChildExecutionInfos
 	e.pendingRequestCancelInfoIDs = state.RequestCancelInfos
 	e.executionInfo = state.ExecutionInfo
+	e.persistedBufferedEvents = state.BufferedEvents
 	for _, ai := range state.ActivitInfos {
 		e.pendingActivityInfoByActivityID[ai.ActivityID] = ai.ScheduleID
 	}
 }
 
-func (e *mutableStateBuilder) CloseUpdateSession() *mutableStateSessionUpdates {
+func (e *mutableStateBuilder) FlushBufferedEvents() error {
+	// put new events into 2 buckets:
+	//  1) if the event was added while there was in-flight decision, then put it in buffered bucket
+	//  2) otherwise, put it in committed bucket
+	var newBufferedEvents []*workflow.HistoryEvent
```

Comment:
I guess this could be simplified as we can never have partial buffered or committed events in the same update.

Reply:
but you still need to check if this is buffered or committed batch.

Reply:
I tried to put the events into buffered or committed based on first event's event id, but it would break lots of unit tests, mostly because we have the tests that addDecisionTaskStarted, then add some other events, then add decisionTaskCompleted. I would spend more time to update those tests, but i feel it is not worth it.
All other comments have been addressed.